### DISCUSSION TITLE
Full Support for WebAuthn 2?

### DISCUSSION THREAD

**Discussion by ohri-anurag** [tweag/webauthn#121] [Q&A] [answered]
Hi,

I came across this library while searching for Haskell versions of the WebAuthn spec. I have following the repo's progress for the past few weeks, I see that it is actively being worked on. Thanks a lot for building this.

I want to know if this library has full support for WebAuthn 2. If not, are there any plans to support the full spec in the future?

Regards,
Anurag Ohri

**Comment by ErinvanderVeen** [comment, marked as answer]
Hi Anurag,

Thanks for taking a look at our library. This library implements almost the entire WebAuthn 2 spec, with two main exceptions:

### tokenBinding
The reasons for which can be found here: https://github.com/tweag/webauthn/blob/7b6e295be028454c7963b1c4fa93f86a99b492db/src/Crypto/WebAuthn/Model/Types.hs#L902 . Copied here for readability:
```haskell
    -- TODO: This library does not implement token binding, this is in
    -- anticipation of version 3 of the webauthn spec that likely removes this
    -- field in its entirety. Discussion can be found in
    -- [the relevant PR](https://github.com/w3c/webauthn/pull/1630).
    -- Chromium and Firefox both don't propagate this field.
    -- Once v3 of the webauthn has been released, and this library is updated
    -- to the new spec, this field and related references can be removed.
```

### Extensions
This is still a feature we would like to have, but currently has a low priority. We are open to contributions from outside of Tweag/Mecury. Discussion of this feature would take place at #35. A relevant comment from within the code is: https://github.com/tweag/webauthn/blob/7b6e295be028454c7963b1c4fa93f86a99b492db/src/Crypto/WebAuthn/Model/Types.hs#L16

### Other
Other than these two, there are several smaller TODOs that are currently not in line with the specification. These are considered low priority because we haven't found an example of an authenticator that requires this part of the spec. For example, the compressed representation of coordinates in ECDSA, the `BiometricStatusReports` in the metadata, and trying `x5u` before checking `x5c` during the decoding of the metadata.

For a basic usage the lack of these features shouldn't hold you back. If they do, we would love to hear from you such that we can improve the library.

We do not consider the library ready for production use yet. We are looking for people/companies willing to test the library such that we can iron out any issues they might encounter.

Kindly,
Erin
